ddtrace/internal/writer.py

```diff
@@ -355,21 +360,29 @@ def recreate(self):
         )
 
     def _put(self, data, headers):
-        conn = get_connection(self.agent_url, self._timeout)
+        # type: (bytes, Dict[str, str]) -> Response
+        if self._conn is None:
+            self._conn = get_connection(self.agent_url, self._timeout)
+
+        headers["Connection"] = "keep-alive"
 
         with StopWatch() as sw:
             try:
-                conn.request("PUT", self._endpoint, data, headers)
-                resp = compat.get_connection_response(conn)
+                self._conn.request("PUT", self._endpoint, data, headers)
+            except Exception:
+                # Close the connection and reset it to None on error.
+                self._conn.close()
+                self._conn = None
+                raise
```

Comment:
Our Tenacity retry logic is going to reattempt on failures right?

My concern is the flow for when `self._conn is not None`, but is in a bad state. We will get an exception, close/None it, then Tenacity will retry?

Are we worried about Tenacity's backoff when this specific case occurs?

Reply:
Yeah the tenacity retry logic will still work like before. When tenacity retries then `self._conn` will be None and we'll recreate the connection - so same as before again.